[temp] feat: add detail taxonomy page

rpenido opened the pull request:
![2023-10-24-185703](https://github.com/openedx/frontend-app-course-authoring/assets/849463/c7335351-8d91-42e2-a4de-04c444a66fe0)
## Supporting Information
- Closes https://github.com/openedx/modular-learning/issues/138
- Depends on:
  - https://github.com/openedx/frontend-app-course-authoring/pull/645
____
Private-ref: [FAL-3529](https://tasks.opencraft.com/browse/FAL-3529)

pomegranited commented on src/index.jsx:
```diff
@@ -71,12 +71,22 @@ const App = () => {
           )}
+          <Route>
+            No match (404)
+          </Route>
```
Is this necessary? I wasn't able to trigger it by visiting a non-existent page, so I think this part should be removed.
```suggestion
```

rpenido replied:
Removed

pomegranited commented on src/taxonomy/api/hooks/api.js:
```diff
@@ -5,10 +5,18 @@ import { getAuthenticatedHttpClient } from '@edx/frontend-platform/auth';
+  getApiBaseUrl(),
+).href;
+const getTagListApiUrl = (taxonomyId) => new URL(
+  `api/content_tagging/v1/taxonomies/${taxonomyId}/tags/?page_size=300`,
```
We shouldn't be hard-coding `page_size` into the API URL here; let the backend determine the default page size.

```suggestion
  `api/content_tagging/v1/taxonomies/${taxonomyId}/tags/`,
```

If we need the page_size to be configurable/modifiable by the user, then we can add it to Paragon's table component.

rpenido replied:
Added backend pagination: 3d6b0dfe53b817861038e2758aa405cc9b182066

pomegranited commented on src/taxonomy/taxonomy-detail/TagListTable.jsx:
```diff
@@ -0,0 +1,49 @@
+      isPaginated
+      isFilterable
+      isSortable
+      defaultColumnValues={{ Filter: TextFilter }}
```
From https://github.com/openedx/modular-learning/issues/138

> Do not implement Keyword filtering for now - that will come in https://github.com/openedx/modular-learning/issues/128

So please remove the filter-related settings here? We'll add them when we're ready for them.

```suggestion
      isSortable
```

pomegranited commented on src/taxonomy/taxonomy-detail/TagListTable.jsx:
```diff
@@ -0,0 +1,49 @@
+
+  return (
+    <DataTable
+      isPaginated
```
Missing a few pagination-related settings here (ref: [Paragon: DataTable](https://paragon-openedx.netlify.app/components/datatable/)):
```suggestion
      isPaginated
      manualPagination
      pageCount={tagList.numPages}
      fetchData={fetchData}
```

^ needs a `fetchData` method implemented to make the next/prev page buttons fetch the next page of data, see the example under [DataTable#paginated-selection](https://paragon-openedx.netlify.app/components/datatable/#backend-filtering-and-sorting).

rpenido replied:
Done 3d6b0dfe53b817861038e2758aa405cc9b182066

pomegranited replied:
I haven't dug into figure out why, but when you hit the prev/next buttons on the table here, we hit both these REST APIs. But we should only be hitting the 2nd `/tags/` endpoint.

1. `/api/content_tagging/v1/taxonomies/<pk>/`
2. `/api/content_tagging/v1/taxonomies/<pk>/tags/`

rpenido replied:
This is also related to `React-Query`. If we refocus the window or [other conditions](https://tanstack.com/query/v4/docs/react/guides/important-defaults), we get a fetch in the background. This is the default behavior. We could tune this application-wise (the same happens in the TaxonomyList) if we think this doesn't suit us.

I was not able to reproduce the issue here. Maybe a windows refocus (not prev/next) triggered the call?

This is from the react-query docs link above:
> If you see a refetch that you are not expecting, it is likely because you just focused the window and TanStack Query is doing a [refetchOnWindowFocus](https://tanstack.com/query/v4/docs/react/guides/window-focus-refetching). During development, this will probably be triggered more frequently, especially because focusing between the Browser DevTools and your app will also cause a fetch, so be aware of that.

bradenmacdonald replied:
Yeah, don't worry about extra fetch calls seen, it's just checking if the server data has changed, so it can automatically refresh the UI if needed. If the data hasn't changed, nothing will happen.

pomegranited commented on src/taxonomy/taxonomy-detail/TaxonomyDetailSideCard.jsx:
```diff
@@ -0,0 +1,27 @@
+    <Card.Section title="Description">
+      {taxonomy.description}
+    </Card.Section>
+    <Card.Divider className="ml-3 mr-3" />
+    <Card.Section title="Copyright">
+      No copyright added
+    </Card.Section>
```
Please omit this until we have copyright fields to show:
```suggestion
```

pomegranited commented on src/taxonomy/taxonomy-card/TaxonomyCard.jsx:
```diff
@@ -101,7 +103,8 @@ const TaxonomyCard = ({ className, original, intl }) => {
   return (
     <>
-      <Card className={classNames('taxonomy-card', className)} data-testid={`taxonomy-card-${id}`}>
+      <Card isClickable className={classNames('taxonomy-card', className)} data-testid={`taxonomy-card-${id}`}>
```
If you remove `isClickable` here, then the stretched-link be what gets followed when you click the card, or when you tab to the card and hit "enter".

rpenido replied:
The `isClickable` adds a little hover effect to the card.
I think the following (591604cfc692e0f20fa734be095f601589c3a2f6) is the right way to do it (without the need of the `.stretched-link` class).

pomegranited commented on src/taxonomy/taxonomy-card/TaxonomyCard.scss:
```diff
@@ -30,6 +30,11 @@
     white-space: nowrap;
   }
 
+  .pgn__card-header-actions, .badge {
+    z-index: 100; 
+    position: relative;
+  }
+
```
This change looks like it belongs on https://github.com/openedx/frontend-app-course-authoring/pull/645, not here?

rpenido replied:
I needed this because the `stretched-link` was over the card menu and the tooltip from the system-defined badge. It was fixed using the clickable card (591604cfc692e0f20fa734be095f601589c3a2f6).

pomegranited reviewed (changes requested):
I've run this on my devstack and it's looking great so far, @rpenido !

I've left a few suggestions, mainly around how the DataTable component is configured.

And as you've noted, there's some copy-paste code in here which should be factored out and moved up under `src/taxonomy` somewhere.

pomegranited commented on src/index.jsx:
```diff
@@ -71,12 +71,22 @@ const App = () => {
+                  );
+                }}
+              />
+            </>
```
There's a routing bug that occurs if you follow these steps:

1. Visit http://localhost:2001/taxonomy-list/   <-- note trailing slash
2. Click on one of the taxonomies there, and note the URL looks like this: `http://localhost:2001/taxonomy-list/taxonomy-list/<pk>`

I'm not sure what the fix is, but I recommend looking at how the other nested routes are written and imitating that.

rpenido replied:
Thank you for getting that!

The react-router was updated (https://github.com/openedx/frontend-app-course-authoring/pull/519) to V6. I will merge this update in the [FAL-3528](https://github.com/open-craft/frontend-app-course-authoring/tree/chris/FAL-3528-taxonomy-export-menu) branch and rebase here before fixing it!

rpenido replied:
Fixed!

pomegranited replied:
Excellent, great fix @rpenido !

And I like how you pulled the common header/footer into a single `TaxonomyLayout` wrapper, that's perfect.

pomegranited commented on src/taxonomy/taxonomy-detail/TagListTable.jsx:
```diff
@@ -0,0 +1,63 @@
+  });
+
+  const useTagListData = () => {
+    const { isError, isFetched, isLoading } = useTagListDataStatus(taxonomyId, options);
+    const tagList = useTagListDataResponse(taxonomyId, options);
```
Calling both `useTagListDataStatus` and `useTagListDataResponse` is unnecessary here -- if you watch the Network tab of your browser Inspector, you can see that we hit the taxonomy REST APIs twice every time we want to load a page of tags.

Both of those methods call the api.js's [`useTagListData`](https://github.com/open-craft/frontend-app-course-authoring/pull/5/files#diff-51e2c0d3363c83c5d167493ee852ee13f16564ad2c70189c74d66f8adaca8cd6R54) (and hit the REST API), they just return different slices of the response object, so it would be better to combine these into a single selector method.

rpenido replied:
Hi, @pomegranited!

This is my first time using `React-Query,` and I was also confused why multiple selectors are recommended by [our docs](https://openedx.atlassian.net/wiki/spaces/BPL/pages/3852632086/Typescript+and+React-Query+for+api+usage.#3%3A-Define-Selectors-for-the-API-hooks), afraid that it will lead to multiple API calls.

But the `React-Query` caches the API call result, so multiple calls are not supposed to happen, even using the current approach.

Could you check if the first call is a pre-flight one (method: `OPTION`)?
![image](https://github.com/open-craft/frontend-app-course-authoring/assets/849463/f91e079c-7ad5-40c9-9914-6e2256545714)
If it is, we are good, and everything works as expected!

bradenmacdonald replied:
Yep, react-query should always be caching the result of each API call, so you can call each endpoint as many times as you want, from different components at different levels, and use different parts of the result, but it should only result in a single API call*. You do have to make sure the URL is exactly the same though, and if there's anything else like user ID that affects the result it should be in the [query key](https://tanstack.com/query/v4/docs/react/guides/query-keys).

* But a single call can be two if you include the preflight, as @rpenido points out.

pomegranited commented on src/taxonomy/taxonomy-detail/TaxonomyDetailPage.jsx:
```diff
@@ -0,0 +1,128 @@
+  const { taxonomyId } = useParams();
+
+  const useTaxonomyDetailData = () => {
+    const { isError, isFetched } = useTaxonomyDetailDataStatus(taxonomyId);
+    const taxonomy = useTaxonomyDetailDataResponse(taxonomyId);
```
Similar comment here as [with the tag list](https://github.com/open-craft/frontend-app-course-authoring/pull/5/files#diff-8e2e37b39ab5b552b68cc89e7dc30ec0789817c2f142fa3bd8d25fa61f8f7e33R17) -- can we avoid hitting the API twice here?

rpenido replied:
Ditto

pomegranited reviewed (changes requested):
These latest changes look great, @rpenido ! I found some issues with the REST API being hit more than we need, but your addition of `TaxonomyLayout`, and your taxonomy-detail components look great.

bradenmacdonald commented:
Here is the PR that will merge soon and affect how taxonomy data is returned from the API - hopefully making it more flexible for you: https://github.com/openedx/edx-platform/pull/33553

pomegranited commented on src/taxonomy/TaxonomyLayout.jsx:
```diff
@@ -0,0 +1,21 @@
+    <style>
+      {`
+        body {
+            background-color: #E9E6E4; /* light-400 */
```
nit: styles belong in `.scss` files?

rpenido replied:
Done: 6a46c29a869bf33eb748b7ef38ab9d0ae5201f38

pomegranited commented on src/taxonomy/taxonomy-card/TaxonomyCardMenu.jsx:
```diff
@@ -26,7 +26,10 @@ const TaxonomyCardMenu = ({
     <>
```
I'm able to use the Export menu item -- clicking it takes me to the Taxonomy detail page.

So I think we need to run `e.preventDefault()` on the `MenuItem` onClick event below too to prevent the click event from bubbling up to the Card. And since there will be more menu items eventually, I think it's best to do it in the `onClickItem` method above, e.g

```javascript
const onClickItem = (e, menuName) => {
    e.preventDefault();
    setMenuIsOpen(false);
    onClickMenuItem(menuName);
  };
```

invoked below like this:

```javascript
<MenuItem className="taxonomy-menu-item" onClick={(e) => onClickItem(e, 'export')}>
```

pomegranited reviewed (approved):
Hi @rpenido there's a bug with the Export menu item, and I noted one optional nit.

But everything else is working great, so I'll approve pending that fix, and then it's ready for upstream review.

:+1:

- [x] I tested this on my devstack.
- [x] I read through the code
- [x] I checked for accessibility issues by using my keyboard to navigate the new and modified components.
- [x] <strike>Includes documentation</strike> N/A
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

xitij2000 commented on src/taxonomy/TaxonomyLayout.scss:
```diff
@@ -0,0 +1,3 @@
+.taxonomy-layout {
+  background-color: #E9E6E4;
```
This is equivalent to bg-light-400, so you can avoid the custom css.

rpenido replied:
Thank you! Fixed here: a12f2621fd4f9288fd0a610aac40143793aabcb2

xitij2000 commented on src/taxonomy/TaxonomyLayout.test.jsx:
```diff
@@ -0,0 +1,48 @@
+    expect(getByTestId('mock-content')).toBeInTheDocument();
+    expect(getByTestId('mock-footer')).toBeInTheDocument();
+  });
+});
```
I don't think this test is adding anything. There is no logic in this component so I think this file can be entirely skipped.

rpenido replied:
It only helps the code coverage. If this metric is not an issue here, let me know, and I remove it!

xitij2000 commented on src/taxonomy/taxonomy-detail/TagListTable.jsx:
```diff
@@ -0,0 +1,49 @@
+  const { tagList } = useTagListData(taxonomyId);
+
+  if (!tagList || !tagList.results) {
+    return 'Loading...';
```
I think you could use the [Paragon Spinner instead](https://paragon-openedx.netlify.app/components/spinner/)

rpenido replied:
This code is outdated. I'm using the Loading status from the DataTable

bradenmacdonald commented:
@rpenido @xitij2000 What's the status of this PR? And how does it relate to https://github.com/openedx/frontend-app-course-authoring/pull/655 ?

rpenido commented:
> @rpenido @xitij2000 What's the status of this PR? And how does it relate to [openedx#655](https://github.com/openedx/frontend-app-course-authoring/pull/655) ?

- The branch of this feature is based on https://github.com/openedx/frontend-app-course-authoring/pull/645 branch.
- The https://github.com/openedx/frontend-app-course-authoring/pull/655 is against `master` and is the official PR.

This is a temp PR against that branch to allow us to review the changes before https://github.com/openedx/frontend-app-course-authoring/pull/645 is merged.

I'm waiting for https://github.com/openedx/frontend-app-course-authoring/pull/645 approval to rebase this branch and bring the improvements from the review to here.

renovate[bot] committed:
fix(deps): update dependency react-transition-group to v4.4.5 (#647)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
fix(deps): update dependency @edx/frontend-component-footer to v12.5.1 (#650)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency @edx/frontend-build to v13.0.4 (#641)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

pomegranited reviewed (approved):
:+1: Thanks for that @rpenido ! This is good for upstream review now. CC @bradenmacdonald 

- [x] I tested this using the REST API on my devstack
- [x] I read through the code
- [x] <strike>I checked for accessibility issues</strike> N/A
- [x] Includes documentation
- [x] Commit structure follows [OEP-0051](https://github.com/edx/open-edx-proposals/pull/189)

KristinAoki committed:
fix: end date error when certificate row not shown (#668)

bradenmacdonald commented:
I guess this is blocked on https://github.com/openedx/frontend-app-course-authoring/pull/645 now ? Once that PR is merged, please open a new PR for this on the upstream repo.

KristinAoki committed:
fix: fidelity typo in preference api (#662)

renovate[bot] committed:
fix(deps): update dependency @edx/frontend-component-header to v4.9.3 (#651)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

rpenido commented:
> I guess this is blocked on [openedx#645](https://github.com/openedx/frontend-app-course-authoring/pull/645) now ? Once that PR is merged, please open a new PR for this on the upstream repo.

Sure!
The PR on upstream is here: https://github.com/openedx/frontend-app-course-authoring/pull/655. I will rebase as soon as we get https://github.com/openedx/frontend-app-course-authoring/pull/645 merged.

renovate[bot] committed:
fix(deps): update dependency @edx/frontend-lib-content-components to v1.175.1 (#663)

* fix(deps): update dependency @edx/frontend-lib-content-components to v1.175.1

* fix: upgrade paragon

---------

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>
Co-authored-by: KristinAoki <kaoki@2u.com>

KristinAoki committed:
fix: modal exit redirect (#659)

KristinAoki committed:
refactor: files-and-videos folder (#672)

zacharis278 committed:
feat: remove old/duplicate proctoring component (#671)

rpenido committed:
feat: add taxonomy detail page

pomegranited commented on src/taxonomy/data/types.mjs:
```diff
@@ -27,6 +27,13 @@
+ * @property {string} format
+ * @property {string} name
+ */
+
```
These changes aren't needed here (artifact of rebase?)

zacharis278 committed:
feat: opt out is not supported by lti proctoring (#673)

This toggle does nothing if an LTI tool is selected. We should hide it in that case.

KristinAoki committed:
fix: total file count update on add and delete (#681)

KristinAoki committed:
fix: combine filter and sort into one modal (#680)

ChrisChV commented on src/taxonomy/tag-list/data/selectors.js:
```diff
@@ -0,0 +1,42 @@
+// @ts-check
```
@rpenido From what I understand from [here](https://github.com/openedx/frontend-app-course-authoring/pull/645#discussion_r1377342836), these would not be considered selectors in the context of the other MFE modules. To avoid colliding with that context is why I have placed functions like this in `apiHooks.js`. But you can wait until https://github.com/openedx/frontend-app-course-authoring/pull/645 is approved to make those changes

ChrisChV commented on src/taxonomy/taxonomy-detail/TaxonomyDetailMenu.jsx:
@rpenido In order to follow the [ADR](https://github.com/openedx/frontend-app-course-authoring/blob/master/docs/decisions/0002-feature-based-application-organization.rst#file-and-directory-naming), we need an `index.js` on this `taxonomy-detail` sub-module for the public interface

ChrisChV reviewed (approved):
@rpenido Looks good :+1: Check my comments, but I think you can wait until https://github.com/openedx/frontend-app-course-authoring/pull/645 is merged to do any change. 

- [X] I tested this: I follow the testing instructions on https://github.com/openedx/frontend-app-course-authoring/pull/655
- [X] I read through the code and considered the security, stability and performance implications of the changes.
- [X] I tested that the UI can be used with a keyboard only (tab order, keyboard controls).
- [X] Includes tests for bugfixes and/or features added.
- [X] Includes documentation

KristinAoki committed:
fix: active transcript preference not loading (#682)

ChrisChV committed:
feat: Taxonomy export menu [FC-0036] (#645)

* feat: System-defined tooltip added
* feat: Taxonomy card menu added. Export menu item added
* feat: Modal for export taxonomy
* feat: Connect with export API
* test: Tests for API and selectors
* feat: Use windows.location.href to call the export endpoint
* test: ExportModal.test added
* style: Delete unnecessary code
* docs: README updated with taxonomy feature
* style: TaxonomyCard updated to a better code style
* style: injectIntl replaced by useIntl on taxonomy pages and components
* refactor: Move and rename taxonomy UI components to match 0002 ADR
* refactor: Move api to data to match with 0002 ADR
* test: Refactor ExportModal tests
* chore: Fix validations
* chore: Lint
* refactor: Moving hooks to apiHooks
* style: Nit on return null

---------

Co-authored-by: Rômulo Penido <romulo@dash.dev.br>
Co-authored-by: Christofer Chavez <christofer@example.com>

KristinAoki committed:
fix: table search filter (#683)

pomegranited commented on src/taxonomy/taxonomy-detail/TaxonomyDetailSideCard.jsx:
```diff
@@ -0,0 +1,32 @@
+        {taxonomy.name}
+      </Card.Section>
+      <Card.Divider className="ml-3 mr-3" />
+      <Card.Section title={intl.formatMessage(messages.taxonomyDetailsHeader)}>
```
nit typo:
```suggestion
      <Card.Section title={intl.formatMessage(messages.taxonomyDetailsDescription)}>
```

renovate[bot] committed:
fix(deps): update dependency @edx/brand to v1.2.3 (#667)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency @edx/frontend-build to v13.0.5 (#676)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

renovate[bot] committed:
chore(deps): update dependency axios-mock-adapter to v1.22.0 (#678)

Co-authored-by: renovate[bot] <29139614+renovate[bot]@users.noreply.github.com>

pomegranited committed:
Merge branch 'master' into rpenido/fal-3529-bare-bones-taxonomy-detail-page

pomegranited committed:
fix: address nits in PR review

pomegranited committed:
refactor: move data/selectors to data/apiHooks

and fix tests to mock useQuery.

pomegranited commented:
Closed in favor of https://github.com/openedx/frontend-app-course-authoring/pull/655
